src/openrct2-ui/windows/InstallTrack.cpp

```diff
@@ -57,408 +57,383 @@ static rct_widget window_install_track_widgets[] = {
+    std::string _trackPath;
+    std::string _trackName;
+    std::vector<uint8_t> _trackDesignPreviewPixels;
+
+public:
+    void SetupTrack(const utf8* path, std::unique_ptr<TrackDesign> trackDesign)
     {
-        context_show_error(STR_UNABLE_TO_LOAD_FILE, STR_NONE, {});
-        return nullptr;
-    }
+        _trackDesign = std::move(trackDesign);
+        _trackPath = path;
+        _trackName = GetNameFromTrackPath(path);
+        _trackDesignPreviewPixels.resize(4 * TRACK_PREVIEW_IMAGE_SIZE);
 
-    object_manager_unload_all_objects();
-    if (_trackDesign->type == RIDE_TYPE_NULL)
-    {
-        log_error("Failed to load track (ride type null): %s", path);
-        return nullptr;
-    }
-    if (object_manager_load_object(&_trackDesign->vehicle_object.Entry) == nullptr)
-    {
-        log_error("Failed to load track (vehicle load fail): %s", path);
-        return nullptr;
+        UpdatePreview();
+        Invalidate();
     }
 
-    window_close_by_class(WindowClass::EditorObjectSelection);
-    window_close_construction_windows();
-
-    gTrackDesignSceneryToggle = false;
-    _currentTrackPieceDirection = 2;
-
-    int32_t screenWidth = context_get_width();
-    int32_t screenHeight = context_get_height();
-    int32_t x = screenWidth / 2 - 201;
-    int32_t y = std::max(TOP_TOOLBAR_HEIGHT + 1, screenHeight / 2 - 200);
-
-    rct_window* w = WindowCreate(ScreenCoordsXY(x, y), WW, WH, &window_install_track_events, WindowClass::InstallTrack, 0);
-    w->widgets = window_install_track_widgets;
-    WindowInitScrollWidgets(*w);
-    w->track_list.track_list_being_updated = false;
-    window_push_others_right(*w);
-
-    _trackPath = path;
-    _trackName = GetNameFromTrackPath(path);
-    _trackDesignPreviewPixels.resize(4 * TRACK_PREVIEW_IMAGE_SIZE);
-
-    WindowInstallTrackUpdatePreview();
-    w->Invalidate();
-
-    return w;
-}
-
-/**
- *
- *  rct2: 0x006D41DC
- */
-static void WindowInstallTrackClose(rct_window* w)
-{
-    _trackPath.clear();
-    _trackName.clear();
-    _trackDesignPreviewPixels.clear();
-    _trackDesignPreviewPixels.shrink_to_fit();
-    _trackDesign = nullptr;
-}
-
-/**
- *
- *  rct2: 0x006D407A
- */
-static void WindowInstallTrackMouseup(rct_window* w, WidgetIndex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnOpen() override
     {
-        case WIDX_CLOSE:
-        case WIDX_CANCEL:
-            window_close(*w);
-            break;
-        case WIDX_ROTATE:
-            _currentTrackPieceDirection++;
-            _currentTrackPieceDirection %= 4;
-            w->Invalidate();
-            break;
-        case WIDX_TOGGLE_SCENERY:
-            gTrackDesignSceneryToggle = !gTrackDesignSceneryToggle;
-            WindowInstallTrackUpdatePreview();
-            w->Invalidate();
-            break;
-        case WIDX_INSTALL:
-            WindowInstallTrackDesign(w);
-            break;
+        widgets = window_install_track_widgets;
+        track_list.track_list_being_updated = false;
```

Comment:
This used to be after `WindowInitScrollWidgets` (not sure if it mattered)

Reply:
From what I understand the order does not matter here.

Reply:
I think it doesn't either.